File: untracked-variant-quantity-drift/python/detect_untracked_drift.py
"""Find Medusa variants with manage_inventory false whose stocked_quantity
has still drifted from a saved baseline (untracked stock that should never
change, but does). Report only, never auto-writes a corrected quantity.

Guide: https://www.allanninal.dev/medusa/untracked-variant-quantity-drift/

Safe to run again and again.
"""
import os
import sys
import json
import logging

import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("detect_untracked_drift")

BASE_URL = os.environ.get("MEDUSA_BACKEND_URL", "http://localhost:9000")
EMAIL = os.environ.get("MEDUSA_ADMIN_EMAIL", "admin@example.com")
PASSWORD = os.environ.get("MEDUSA_ADMIN_PASSWORD", "supersecret")
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"
BASELINE_PATH = os.environ.get("BASELINE_PATH", "untracked_drift_baseline.json")

PRODUCT_FIELDS = (
    "id,title,*variants,variants.manage_inventory,"
    "variants.inventory_items.inventory.id,"
    "variants.inventory_items.inventory.location_levels.stocked_quantity,"
    "variants.inventory_items.inventory.location_levels.reserved_quantity,"
    "variants.inventory_items.inventory.location_levels.location_id"
)


def get_token():
    r = requests.post(
        f"{BASE_URL}/auth/user/emailpass",
        json={"email": EMAIL, "password": PASSWORD},
        timeout=30,
    )
    r.raise_for_status()
    return r.json()["token"]


def list_products(token):
    headers = {"Authorization": f"Bearer {token}"}
    out, offset, limit = [], 0, 100
    while True:
        r = requests.get(
            f"{BASE_URL}/admin/products",
            params={"fields": PRODUCT_FIELDS, "limit": limit, "offset": offset},
            headers=headers,
            timeout=30,
        )
        r.raise_for_status()
        body = r.json()
        out.extend(body["products"])
        offset += limit
        if offset >= body["count"]:
            return out


def flatten_variants(products):
    """Reshape raw product/variant payloads into the plain records the pure
    decision function expects."""
    flat = []
    for product in products:
        for variant in product.get("variants") or []:
            inventory_items = variant.get("inventory_items") or []
            first_item = inventory_items[0]["inventory"] if inventory_items else None
            item_id = first_item["id"] if first_item else None
            levels = []
            if first_item:
                for lvl in first_item.get("location_levels") or []:
                    levels.append({
                        "locationId": lvl["location_id"],
                        "stockedQuantity": lvl["stocked_quantity"],
                    })
            flat.append({
                "variantId": variant["id"],
                "sku": variant.get("sku"),
                "productTitle": product.get("title"),
                "manageInventory": bool(variant.get("manage_inventory")),
                "inventoryItemId": item_id,
                "locationLevels": levels,
            })
    return flat


def load_baseline(path):
    try:
        with open(path) as f:
            raw = json.load(f)
    except FileNotFoundError:
        return {}
    return {item_id: dict(locs) for item_id, locs in raw.items()}


def save_baseline(path, variants):
    snapshot = {}
    for variant in variants:
        item_id = variant.get("inventoryItemId")
        if not item_id:
            continue
        locs = snapshot.setdefault(item_id, {})
        for level in variant.get("locationLevels") or []:
            locs[level["locationId"]] = level["stockedQuantity"]
    with open(path, "w") as f:
        json.dump(snapshot, f, indent=2, sort_keys=True)


def detect_untracked_quantity_drift(variants, baseline):
    """Pure: no I/O. variants is a list of {variantId, manageInventory,
    inventoryItemId, locationLevels: [{locationId, stockedQuantity}]}.
    baseline is inventoryItemId -> locationId -> lastKnownStockedQuantity.

    Skips tracked variants (manage_inventory true), skips variants with no
    linked inventory item or no location levels, and only reports a record
    when the delta between current and baseline quantity is nonzero, since
    any change at all on a supposedly untracked variant is suspect.
    """
    drifted = []
    for variant in variants:
        if variant.get("manageInventory"):
            continue
        item_id = variant.get("inventoryItemId")
        levels = variant.get("locationLevels") or []
        if not item_id or not levels:
            continue

        item_baseline = baseline.get(item_id) or {}
        for level in levels:
            location_id = level["locationId"]
            current = level["stockedQuantity"]
            if location_id not in item_baseline:
                continue
            base_qty = item_baseline[location_id]
            delta = current - base_qty
            if delta != 0:
                drifted.append({
                    "variantId": variant["variantId"],
                    "inventoryItemId": item_id,
                    "locationId": location_id,
                    "baselineQuantity": base_qty,
                    "currentQuantity": current,
                    "delta": delta,
                })
    return drifted


def restore_baseline_quantity(token, inventory_item_id, location_id, baseline_quantity):
    headers = {"Authorization": f"Bearer {token}"}
    r = requests.post(
        f"{BASE_URL}/admin/inventory-items/{inventory_item_id}/location-levels/{location_id}",
        json={"stocked_quantity": baseline_quantity},
        headers=headers,
        timeout=30,
    )
    r.raise_for_status()
    return r.json()


def run():
    # An operator can pass --restore=VARIANT_ID=QTY pairs to approve a
    # specific restore. Nothing is ever inferred automatically.
    restore_map = {}
    for arg in sys.argv[1:]:
        if arg.startswith("--restore="):
            pair = arg[len("--restore="):]
            variant_id, qty = pair.split("=")
            restore_map[variant_id] = int(qty)

    token = get_token()
    products = list_products(token)
    variants = flatten_variants(products)

    baseline = load_baseline(BASELINE_PATH)
    drift = detect_untracked_quantity_drift(variants, baseline)

    if not drift:
        log.info("No drift found across %d variant(s).", len(variants))
    else:
        for record in drift:
            log.warning(
                "Drift: variant %s, inventory_item %s, location %s. "
                "baseline=%s current=%s delta=%s",
                record["variantId"], record["inventoryItemId"], record["locationId"],
                record["baselineQuantity"], record["currentQuantity"], record["delta"],
            )
        log.info("Done. %d drifted record(s) found.", len(drift))

    if not DRY_RUN and restore_map:
        by_variant = {v["variantId"]: v for v in variants}
        for variant_id, target_qty in restore_map.items():
            variant = by_variant.get(variant_id)
            if not variant or not variant["inventoryItemId"]:
                log.warning("Skipping restore for %s, variant or inventory item not found.", variant_id)
                continue
            for level in variant["locationLevels"]:
                log.info(
                    "Restoring variant %s location %s from %s to operator-confirmed %s.",
                    variant_id, level["locationId"], level["stockedQuantity"], target_qty,
                )
                restore_baseline_quantity(token, variant["inventoryItemId"], level["locationId"], target_qty)

    save_baseline(BASELINE_PATH, variants)


if __name__ == "__main__":
    run()

File: untracked-variant-quantity-drift/node/detect-untracked-drift.js
/**
 * Find Medusa variants with manage_inventory false whose stocked_quantity has
 * still drifted from a saved baseline (untracked stock that should never
 * change, but does). Report only, never auto-writes a corrected quantity.
 *
 * Guide: https://www.allanninal.dev/medusa/untracked-variant-quantity-drift/
 *
 * Safe to run again and again.
 */
import { pathToFileURL } from "node:url";
import { readFileSync, writeFileSync } from "node:fs";

const BASE_URL = process.env.MEDUSA_BACKEND_URL || "http://localhost:9000";
const EMAIL = process.env.MEDUSA_ADMIN_EMAIL || "admin@example.com";
const PASSWORD = process.env.MEDUSA_ADMIN_PASSWORD || "supersecret";
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";
const BASELINE_PATH = process.env.BASELINE_PATH || "untracked_drift_baseline.json";

const PRODUCT_FIELDS = [
  "id,title,*variants,variants.manage_inventory,",
  "variants.inventory_items.inventory.id,",
  "variants.inventory_items.inventory.location_levels.stocked_quantity,",
  "variants.inventory_items.inventory.location_levels.reserved_quantity,",
  "variants.inventory_items.inventory.location_levels.location_id",
].join("");

export function detectUntrackedQuantityDrift(variants, baseline) {
  // Pure: no I/O. variants is [{ variantId, manageInventory, inventoryItemId,
  // locationLevels: [{ locationId, stockedQuantity }] }]. baseline is a Map of
  // inventoryItemId -> Map of locationId -> lastKnownStockedQuantity.
  //
  // Skips tracked variants (manageInventory true), skips variants with no
  // linked inventory item or no location levels, and only reports a record
  // when the delta between current and baseline quantity is nonzero, since
  // any change at all on a supposedly untracked variant is suspect.
  const drifted = [];
  for (const variant of variants) {
    if (variant.manageInventory) continue;
    const itemId = variant.inventoryItemId;
    const levels = variant.locationLevels || [];
    if (!itemId || levels.length === 0) continue;

    const itemBaseline = baseline.get(itemId);
    if (!itemBaseline) continue;

    for (const level of levels) {
      const { locationId, stockedQuantity: current } = level;
      if (!itemBaseline.has(locationId)) continue;
      const baselineQuantity = itemBaseline.get(locationId);
      const delta = current - baselineQuantity;
      if (delta !== 0) {
        drifted.push({
          variantId: variant.variantId,
          inventoryItemId: itemId,
          locationId,
          baselineQuantity,
          currentQuantity: current,
          delta,
        });
      }
    }
  }
  return drifted;
}

async function login() {
  const { default: Medusa } = await import("@medusajs/js-sdk");
  const sdk = new Medusa({ baseUrl: BASE_URL, auth: { type: "jwt" } });
  await sdk.auth.login("user", "emailpass", { email: EMAIL, password: PASSWORD });
  return sdk;
}

async function listProducts(sdk) {
  const out = [];
  let offset = 0;
  const limit = 100;
  while (true) {
    const body = await sdk.admin.product.list({ fields: PRODUCT_FIELDS, limit, offset });
    out.push(...body.products);
    offset += limit;
    if (offset >= body.count) return out;
  }
}

function flattenVariants(products) {
  const flat = [];
  for (const product of products) {
    for (const variant of product.variants || []) {
      const inventoryItems = variant.inventory_items || [];
      const firstItem = inventoryItems[0]?.inventory || null;
      const itemId = firstItem?.id || null;
      const levels = (firstItem?.location_levels || []).map((lvl) => ({
        locationId: lvl.location_id,
        stockedQuantity: lvl.stocked_quantity,
      }));
      flat.push({
        variantId: variant.id,
        sku: variant.sku,
        productTitle: product.title,
        manageInventory: Boolean(variant.manage_inventory),
        inventoryItemId: itemId,
        locationLevels: levels,
      });
    }
  }
  return flat;
}

function loadBaseline(path) {
  let raw;
  try {
    raw = JSON.parse(readFileSync(path, "utf8"));
  } catch {
    return new Map();
  }
  const baseline = new Map();
  for (const [itemId, locs] of Object.entries(raw)) {
    baseline.set(itemId, new Map(Object.entries(locs)));
  }
  return baseline;
}

function saveBaseline(path, variants) {
  const snapshot = {};
  for (const variant of variants) {
    const itemId = variant.inventoryItemId;
    if (!itemId) continue;
    const locs = (snapshot[itemId] ||= {});
    for (const level of variant.locationLevels || []) {
      locs[level.locationId] = level.stockedQuantity;
    }
  }
  writeFileSync(path, JSON.stringify(snapshot, null, 2));
}

async function restoreBaselineQuantity(sdk, inventoryItemId, locationId, baselineQuantity) {
  return sdk.admin.inventoryItem.updateLocationLevel(inventoryItemId, locationId, {
    stocked_quantity: baselineQuantity,
  });
}

export async function run() {
  // An operator can pass --restore=VARIANT_ID=QTY to approve a specific
  // restore. Nothing is ever inferred automatically.
  const restoreMap = new Map();
  for (const arg of process.argv.slice(2)) {
    if (arg.startsWith("--restore=")) {
      const pair = arg.slice("--restore=".length);
      const [variantId, qty] = pair.split("=");
      restoreMap.set(variantId, Number(qty));
    }
  }

  const sdk = await login();
  const products = await listProducts(sdk);
  const variants = flattenVariants(products);

  const baseline = loadBaseline(BASELINE_PATH);
  const drift = detectUntrackedQuantityDrift(variants, baseline);

  if (drift.length === 0) {
    console.log(`No drift found across ${variants.length} variant(s).`);
  } else {
    for (const record of drift) {
      console.warn(
        `Drift: variant ${record.variantId}, inventory_item ${record.inventoryItemId}, ` +
        `location ${record.locationId}. baseline=${record.baselineQuantity} ` +
        `current=${record.currentQuantity} delta=${record.delta}`
      );
    }
    console.log(`Done. ${drift.length} drifted record(s) found.`);
  }

  if (!DRY_RUN && restoreMap.size > 0) {
    const byVariant = new Map(variants.map((v) => [v.variantId, v]));
    for (const [variantId, targetQty] of restoreMap) {
      const variant = byVariant.get(variantId);
      if (!variant || !variant.inventoryItemId) {
        console.warn(`Skipping restore for ${variantId}, variant or inventory item not found.`);
        continue;
      }
      for (const level of variant.locationLevels) {
        console.log(
          `Restoring variant ${variantId} location ${level.locationId} from ` +
          `${level.stockedQuantity} to operator-confirmed ${targetQty}.`
        );
        await restoreBaselineQuantity(sdk, variant.inventoryItemId, level.locationId, targetQty);
      }
    }
  }

  saveBaseline(BASELINE_PATH, variants);
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}
